File: js/servewright-react/src/transition.ts
import type { ServewrightNode, Transition, View } from "./types.js";

export type Patch =
  | { op: "replace"; target: string; node: ServewrightNode }
  | { op: "insert"; parent: string; index: number; node: ServewrightNode }
  | { op: "remove"; target: string }
  | { op: "setError"; target: string; errors: string[] }
  | { op: "setLoading"; target: string; loading: boolean };

export class TransitionDesyncError extends Error {
  constructor(expectedBasedOn: number, actualStateVersion: number) {
    super(`Transition basedOn=${expectedBasedOn} does not match client stateVersion=${actualStateVersion}`);
    this.name = "TransitionDesyncError";
  }
}

export function applyTransition(
  view: View,
  transition: Transition,
  dirtyFields: Set<string> = new Set(),
): View {
  if (transition.basedOn !== view.stateVersion) {
    throw new TransitionDesyncError(transition.basedOn, view.stateVersion);
  }

  let root = view.root;
  for (const patch of transition.patches as Patch[]) {
    root = applyPatch(root, patch, dirtyFields);
  }

  return {
    ...view,
    stateVersion: transition.stateVersion,
    root,
  };
}

function applyPatch(node: ServewrightNode, patch: Patch, dirtyFields: Set<string>): ServewrightNode {
  switch (patch.op) {
    case "replace":
      if (node.id === patch.target) {
        if (dirtyFields.has(patch.target) && isInputNode(patch.node)) {
          return preserveLocalValue(patch.node, node);
        }
        return patch.node;
      }
      return {
        ...node,
        children: node.children?.map((child) => applyPatch(child, patch, dirtyFields)),
      };
    case "insert":
      if (node.id === patch.parent) {
        const children = [...(node.children ?? [])];
        children.splice(patch.index, 0, patch.node);
        return { ...node, children };
      }
      return {
        ...node,
        children: node.children?.map((child) => applyPatch(child, patch, dirtyFields)),
      };
    case "remove":
      if (node.id === patch.target) {
        return node;
      }
      return {
        ...node,
        children: node.children
          ?.filter((child) => child.id !== patch.target)
          .map((child) => applyPatch(child, patch, dirtyFields)),
      };
    case "setError":
      if (node.id === patch.target) {
        return {
          ...node,
          props: {
            ...node.props,
            errors: patch.errors,
          },
        };
      }
      return {
        ...node,
        children: node.children?.map((child) => applyPatch(child, patch, dirtyFields)),
      };
    case "setLoading":
      if (node.id === patch.target) {
        const props = { ...node.props };
        if (patch.loading) {
          props.loading = true;
        } else {
          delete props.loading;
        }
        return { ...node, props };
      }
      return {
        ...node,
        children: node.children?.map((child) => applyPatch(child, patch, dirtyFields)),
      };
  }
}

function isInputNode(node: ServewrightNode): boolean {
  return node.type === "TextInput" || node.type === "Select" || node.type === "Checkbox";
}

function preserveLocalValue(incoming: ServewrightNode, current: ServewrightNode): ServewrightNode {
  if (!isInputNode(incoming) || !isInputNode(current)) {
    return incoming;
  }
  return {
    ...incoming,
    props: {
      ...incoming.props,
      value: current.props.value,
    },
  };
}

export function collectDirtyFields(
  view: View,
  localValues: Record<string, string>,
): Set<string> {
  const dirty = new Set<string>();
  for (const [fieldId, localValue] of Object.entries(localValues)) {
    const serverValue = findNodeById(view.root, fieldId)?.props.value;
    if (String(serverValue ?? "") !== localValue) {
      dirty.add(fieldId);
    }
  }
  return dirty;
}

function findNodeById(node: ServewrightNode, id: string): ServewrightNode | undefined {
  if (node.id === id) {
    return node;
  }
  for (const child of node.children ?? []) {
    const found = findNodeById(child, id);
    if (found) {
      return found;
    }
  }
  return undefined;
}

File: dart/servewright_flutter/lib/src/transition.dart
import 'package:servewright_flutter/src/types.dart';

class TransitionDesyncError implements Exception {
  TransitionDesyncError(this.expectedBasedOn, this.actualStateVersion);

  final int expectedBasedOn;
  final int actualStateVersion;

  @override
  String toString() =>
      'Transition basedOn=$expectedBasedOn does not match client stateVersion=$actualStateVersion';
}

class ServewrightTransition {
  const ServewrightTransition({
    required this.basedOn,
    required this.stateVersion,
    required this.patches,
  });

  final int basedOn;
  final int stateVersion;
  final List<Map<String, dynamic>> patches;

  factory ServewrightTransition.fromJson(Map<String, dynamic> json) {
    return ServewrightTransition(
      basedOn: json['basedOn'] as int,
      stateVersion: json['stateVersion'] as int,
      patches: (json['patches'] as List<dynamic>).cast<Map<String, dynamic>>(),
    );
  }
}

class TransitionApplier {
  static ServewrightView apply(
    ServewrightView view,
    ServewrightTransition transition, {
    Set<String> dirtyFields = const {},
  }) {
    if (transition.basedOn != view.stateVersion) {
      throw TransitionDesyncError(transition.basedOn, view.stateVersion);
    }

    var root = view.root;
    for (final patch in transition.patches) {
      root = _applyPatch(root, patch, dirtyFields);
    }

    return ServewrightView(
      servewrightVersion: view.servewrightVersion,
      schemaVersion: view.schemaVersion,
      screen: view.screen,
      stateVersion: transition.stateVersion,
      root: root,
    );
  }

  static Set<String> collectDirtyFields(
    ServewrightView view,
    Map<String, String> localValues,
  ) {
    final dirty = <String>{};
    for (final entry in localValues.entries) {
      final node = _findNodeById(view.root, entry.key);
      final serverValue = '${node?.props['value'] ?? ''}';
      if (serverValue != entry.value) {
        dirty.add(entry.key);
      }
    }
    return dirty;
  }

  static ServewrightNode _applyPatch(
    ServewrightNode node,
    Map<String, dynamic> patch,
    Set<String> dirtyFields,
  ) {
    final op = patch['op'] as String;
    switch (op) {
      case 'replace':
        final target = patch['target'] as String;
        final replacement = ServewrightNode.fromJson(patch['node'] as Map<String, dynamic>);
        if (node.id == target) {
          if (dirtyFields.contains(target) && _isInput(replacement)) {
            return ServewrightNode(
              id: replacement.id,
              type: replacement.type,
              props: {...replacement.props, 'value': node.props['value']},
              children: replacement.children,
            );
          }
          return replacement;
        }
        return ServewrightNode(
          id: node.id,
          type: node.type,
          props: node.props,
          children: node.children
              .map((child) => _applyPatch(child, patch, dirtyFields))
              .toList(),
        );
      case 'setError':
        final target = patch['target'] as String;
        if (node.id == target) {
          return ServewrightNode(
            id: node.id,
            type: node.type,
            props: {...node.props, 'errors': patch['errors']},
            children: node.children,
          );
        }
        return ServewrightNode(
          id: node.id,
          type: node.type,
          props: node.props,
          children: node.children
              .map((child) => _applyPatch(child, patch, dirtyFields))
              .toList(),
        );
      case 'setLoading':
        final target = patch['target'] as String;
        if (node.id == target) {
          final props = Map<String, dynamic>.from(node.props);
          if (patch['loading'] == true) {
            props['loading'] = true;
          } else {
            props.remove('loading');
          }
          return ServewrightNode(id: node.id, type: node.type, props: props, children: node.children);
        }
        return ServewrightNode(
          id: node.id,
          type: node.type,
          props: node.props,
          children: node.children
              .map((child) => _applyPatch(child, patch, dirtyFields))
              .toList(),
        );
      case 'insert':
        final parent = patch['parent'] as String;
        final index = patch['index'] as int;
        final inserted = ServewrightNode.fromJson(patch['node'] as Map<String, dynamic>);
        if (node.id == parent) {
          final children = List<ServewrightNode>.from(node.children);
          children.insert(index, inserted);
          return ServewrightNode(
            id: node.id,
            type: node.type,
            props: node.props,
            children: children,
          );
        }
        return ServewrightNode(
          id: node.id,
          type: node.type,
          props: node.props,
          children: node.children
              .map((child) => _applyPatch(child, patch, dirtyFields))
              .toList(),
        );
      case 'remove':
        final target = patch['target'] as String;
        if (node.id == target) {
          return node;
        }
        return ServewrightNode(
          id: node.id,
          type: node.type,
          props: node.props,
          children: node.children
              .where((child) => child.id != target)
              .map((child) => _applyPatch(child, patch, dirtyFields))
              .toList(),
        );
      default:
        return node;
    }
  }

  static bool _isInput(ServewrightNode node) =>
      node.type == 'TextInput' || node.type == 'Select' || node.type == 'Checkbox';

  static ServewrightNode? _findNodeById(ServewrightNode node, String id) {
    if (node.id == id) {
      return node;
    }
    for (final child in node.children) {
      final found = _findNodeById(child, id);
      if (found != null) {
        return found;
      }
    }
    return null;
  }
}
